keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,952 @@
+  field as nodes are cluster level resources.
+
+## Design Details
+
+### ClusterNetworkPolicy API Design
+The following new `ClusterNetworkPolicy` API will be added to the `networking.k8s.io` API group:
+
+```golang
+type ClusterNetworkPolicy struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec ClusterNetworkPolicySpec
+}
+
+type ClusterNetworkPolicySpec struct {
+	// No implicit isolation of AppliedTo Pods.
+	AppliedTo    AppliedTo
+	Ingress      []ClusterNetworkPolicyIngressRule
+	Egress       []ClusterNetworkPolicyEgressRule
+}
+
+type ClusterNetworkPolicyIngress/EgressRule struct {
+	Action       RuleAction
+	Ports        []networkingv1.NetworkPolicyPort
+	From/To      []ClusterNetworkPolicyPeer
+}
+
+type ClusterNetworkPolicyPeer struct {
+	PodSelector  *metav1.LabelSelector
+	// required if a PodSelector is specified
+	Namespaces   *Namespaces
+	IPBlock      *networkingv1.IPBlock
+}
+
+type ClusterNetworkPolicyExcept struct {
+	PodSelector  *metav1.LabelSelector
+	Namespaces   *Namespaces
+}
+
+const (
+	RuleActionDeny  RuleAction = "Deny"
+	RuleActionAllow RuleAction = "Allow"
+)
+```
+
+For the ClusterNetworkPolicy ingress/egress rule, the `Action` field dictates whether
+traffic should be allowed or denied from/to the ClusterNetworkPolicyPeer.
+This will be a required field.
+
+### DefaultNetworkPolicy API Design
+The following new `DefaultNetworkPolicy` API will be added to the `networking.k8s.io` API group:
+
+```golang
+type DefaultNetworkPolicy struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec DefaultNetworkPolicySpec
+}
+
+type DefaultNetworkPolicySpec struct {
+	// Implicit isolation of AppliedTo Pods.
+	AppliedTo   AppliedTo
+	Ingress     []DefaultNetworkPolicyIngressRule
+	Egress      []DefaultNetworkPolicyEgressRule
+}
+
+type DefaultNetworkPolicyIngress/EgressRule struct {
+	Ports            []networkingv1.NetworkPolicyPort
+	OnlyFrom/OnlyTo  []DefaultNetworkPolicyPeer
+}
+
+type DefaultNetworkPolicyPeer struct {
+	PodSelector  *metav1.LabelSelector
+	// required if a PodSelector is specified
+	Namespaces   *Namespaces
+	IPBlock      *networkingv1.IPBlock
+}
+```
+
+Most structs above are very similar to NetworkPolicy and quite self-explanatory.
+One detail to notice is that in the DefaultNetworkPolicy Ingress/Egress rule spec,
+the peers are created in a field named `OnlyFrom`/`OnlyTo`, as opposed to `To`/`From`
+in ClusterNetworkPolicy. We chose this naming to better hint policy writers about
+the isolation effect of DefaultNetworkPolicy on the workloads it applies to.
+
+### Shared API Design
+The following structs will be added to the `networking.k8s.io` API group and
+shared between `ClusterNetworkPolicy` and `DefaultNetworkPolicy`:
+
+```golang
+type AppliedTo struct {
+	// required if a PodSelector is specified
+	NamespaceSelector   *metav1.LabelSelector
+	// optional
+	PodSelector         *metav1.LabelSelector
+}
+
+type Namespaces struct {
+	// Self and Selector are mutually-exclusive
+	Self       bool
```

Comment:
Although both `self` and `selector` are about namespace, they are not very logically related. Would it be simpler to make them flat in `ClusterNetworkPolicyPeer`? I mean something like:
```
type ClusterNetworkPolicyPeer struct {
	PodSelector *metav1.LabelSelector

	// Select peer pods from namespaces selected by the namespaceSelector.
	NamespaceSelector *metav1.LabelSelector
        // Select peer pods from their own namespaces of the Pods which the ingress/egress rule is currently being evaluated upon.
	SelfNamespace bool

	IPBlock *networkingv1.IPBlock
}
``` 
Users who have been familar with `podSelector` and `namespaceSelector` can continue to define peer in the way they used to, and it's also easy to try out the new `selfNamespace` field in conjunction with `podSelector`.

Reply:
this is a good idea.. will propose this in the next meeting